### PR TITLE
Comparison operators

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libsolidity/semanticTests/viaYul/comparison.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    function neq(uint a, uint b) public pure returns (bool) {
+        return a != b;
+    }
+}
+// ====
+// compileViaYul: true
+// ----
+// f(address): 0x1234 -> false
+// f(address): 0x0 -> true
```
Comment (erak) [inline test/libsolidity/semanticTests/viaYul/comparison.sol]: This breaks the test. Back then we decided to not allow odd hex-literals.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -188,18 +188,59 @@ bool IRGeneratorForStatements::visit(Return const& _return)
 	return false;
 }
 
+void IRGeneratorForStatements::endVisit(UnaryOperation const& _unaryOperation)
+{
+	if (_unaryOperation.annotation().type->category() == Type::Category::RationalNumber)
```
Comment (erak) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Perhaps declare a local variable `auto const* type = _unaryOperation.annotation().type`? But it also fine like this..

Comment (erak) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Ah, and maybe  `solAssert(_unaryOperation.annotation().type, "")`

**Comment by erak** [comment]
Addressing the comment on the test should fix it. Otherwise it looks good already.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `b4607d7be387ae21a99c68c9c4f2abe8cb069e96`:
```
Error: Trailing whitespace found:
 test/libsolidity/semanticTests/viaYul/string_literals.sol:18:// short_dyn(): -> 
 test/libsolidity/semanticTests/viaYul/string_literals.sol:19:// long_dyn(): -> 

```
Please check that your changes are working as intended.

File: test/libsolidity/semanticTests/viaYul/comparison.sol (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/viaYul/comparison.sol]: Fixed.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This is something the type checker should guarantee, but perhaps I can add a function `type(Expression const&)` that always checks it.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `8f569e646b4003672d4c6539b545c3d226159bba`:
```
Error: Trailing whitespace found:
 test/libsolidity/semanticTests/viaYul/string_literals.sol:18:// short_dyn(): -> 
 test/libsolidity/semanticTests/viaYul/string_literals.sol:19:// long_dyn(): -> 

```
Please check that your changes are working as intended.

**Comment by leonardoalt** [comment]
Please rebase

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**